src/supported-operations/running-monitoring-operations.md

```diff
@@ -0,0 +1,94 @@
+  * `"status": "failed"`: when the operation returns with a non-zero status.
+  * NOTE: One might also consider `cancelled` and `aborted` operations to deal with user cancellation
+    and operations killed by a signal or a power failure.
+* The mapper might subscribe to the `tedge/operations/status` topic
+  and translate the operation status into cloud specific messages.
+  * NOTE: There is open question here.
+    * Does the c8y mapper translate the `tedge/operations/status` messages into equivalent smartRest2 messages?
+    * Or does the c8y operation publish their progress directly to `c8y/s/us`?
+    * The former would be simpler for the operator implementor. 
+      However, most operations will in any case have to dialog over `c8y/s/us` and `c8y/s/ds`
+      to implement the specificities of the operation protocol.
+* The execution of a command is logged in file name after the operation name and the request id.
+  * `/var/log/tedge/agent/{operation-name}-{operation-request-id}-{starting-datetime}.log`
+  * These logs are build as for the sm plugin processes
+    (see [logged_command.rs](https://github.com/thin-edge/thin-edge.io/blob/main/crates/core/plugin_sm/src/logged_command.rs))
+    with 4 points.
+    * The command path and the parameters.
+    * The exit status.  
+    * The stdout of the process.
+    * The stderr of the process.
+* The `tedge/operations/status` topic and `/var/log/tedge/agent` logs are used to monitor the operations on a device.
+  The agent also maintains a journal used to recover the previous state on restart.
+  * This journal is actionable, i.e. contains the information
+    to execute a scheduled operation or to stop a running process.
+  * Each time an operation is scheduled, launched, completed,
+    the agent logs this state in its journal with all the information to resume that state and proceed to the next step.
+  * On restart, the agent checks the journal for operations to start or cancel.
+    * An operation that was scheduled is rescheduled.
+    * An operation that was running is marked as aborted.
+    * NOTE: can we check that a detached process is still running (comparing pid or program name)?
+  * NOTE: A key open question is: what mechanism to implement this journal?
+    * The requirements are:
+      * Atomic append of an entry.
+      * Crash-safe.  
+      * Sequential read on restart.
+      * Old entries removal.
+    * The alternatives are:
+      * An append-only text file. Can be acceptable for a first drop, but will stay fragile.
+      * Sqlite fills the bill and much more but also introduces a new external dependency.
+      * An embedded database as [`sled`](https://github.com/spacejam/sled) is closer to our requirements,
+        but might be a bit heavy for an embedded device. 
+      * A commit log as [`commitlog`](https://lib.rs/crates/commitlog) can be the perfect fit.
+        It raises questions about adoption/support though.
```

Comment:
As much as I would go straight in with `sqlite` as it is a gold standard for embedded databases I can see why we wouldn't want to introduce it. Footprint wise is not much of a burden as it may be somewhere in a few hundred KBs which is less than most of our executables atm. Dependency wise it's true that it adds some more to manage and I think that would be the biggest drawback on using using (beside having to use SQL to read/write).
I think plain file would be a good option to start with, however, I am concerned with the amount of maintenance and additional code required to keep it correct. Plus, this option may be missing important persistence mechanism unless we implement our own.
I think `sled` is a great choice and in between above mentioned as it provides great performance (if we need it), straightforward API and serde utilization for data store. As I agree it may add "significant" footprint so would using an external dependency, but in this case as it is Rust package we have a control over how it is used.

My vote goes for `sled`.

Reply:
And why not [rustbreak](https://github.com/TheNeikos/rustbreak) since Marcel in onboard now!

Reply:
I do not have any idea about scale of operations that are handled by an edge(thin-edge) device for a given period of time. Also, the logs are kept only for the last few operations right?. From the explanation, it looks like it's very important for the recovery process in case of any failures. So, do we really need a database here?

Reply:
The load will be low. This is to log pending operations issued, on the device, for maintenance purpose, by an operator.
Robustness is key. The operations must be launched with reliability.